core/src/main/java/com/datastrato/gravitino/authorization/AuthorizationManager.java

```diff
@@ -0,0 +1,261 @@
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Streams;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.io.Closeable;
+import java.io.File;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.ServiceLoader;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Manages the authorization instances and operations. */
+public class AuthorizationManager implements Closeable {
+  private static final Logger LOG = LoggerFactory.getLogger(AuthorizationManager.class);
+
+  public static final String AUTHORIZATION_PROVIDER = "AUTHORIZATION_PROVIDER";
+
+  private final Config config;
+
+  @VisibleForTesting final Cache<CatalogEntity, AuthorizationWrapper> authorizationCache;
+
+  public AuthorizationManager(Config config) {
+    this.config = config;
+
+    long cacheEvictionIntervalInMs = config.get(Configs.AUTHORIZATION_CACHE_EVICTION_INTERVAL_MS);
+    this.authorizationCache =
+        Caffeine.newBuilder()
+            .expireAfterAccess(cacheEvictionIntervalInMs, TimeUnit.MILLISECONDS)
+            .removalListener(
+                (k, v, c) -> {
+                  LOG.info("Closing authorization {}.", k);
+                  ((AuthorizationWrapper) v).close();
+                })
+            .scheduler(
+                Scheduler.forScheduledExecutorService(
+                    new ScheduledThreadPoolExecutor(
+                        1,
+                        new ThreadFactoryBuilder()
+                            .setDaemon(true)
+                            .setNameFormat("authorization-cleaner-%d")
+                            .build())))
+            .build();
+  }
+
+  @Override
+  public void close() {
+    authorizationCache.invalidateAll();
+  }
+
+  public AuthorizationWrapper loadAuthorizationAndWrap(CatalogEntity entity)
+      throws NoSuchAuthorizationException {
+    return authorizationCache.get(entity, this::createAuthorizationWrapper);
+  }
+
+  /** Wrapper class for an Authorization instance and its class loader. */
+  public static class AuthorizationWrapper {
+    private BaseAuthorization authorization;
+
+    private IsolatedClassLoader classLoader;
+
+    public AuthorizationWrapper(BaseAuthorization auth, IsolatedClassLoader classLoader) {
+      this.authorization = auth;
+      this.classLoader = classLoader;
+    }
+
+    public <R> boolean runAuthorizationChain(Function<AuthorizationOperations, R>... functions) {
+      try {
+        return classLoader.withClassLoader(
+            cl -> {
+              for (Function<AuthorizationOperations, R> function : functions) {
```

Comment:
So, we can't guarantee we run `runAuthorizationChain` transactionally?

Reply:
If we manage authorizations in a transactional way, it would be too complicated.
So I'm considering using an idempotent approach to authorization processing, where the result is the same no matter how many times this authorization operation is repeated. This avoids the complexity of using transactions.